plaidml/bridge/openvino/ops/detection_output.cpp

```diff
@@ -0,0 +1,223 @@
+          DecodeBoxes(priors, variances, arm_loc, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+      decoded_bboxes =
+          DecodeBoxes(decoded_priors, variances, location, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+    } else {
+      decoded_bboxes =
+          DecodeBoxes(priors, variances, location, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+    }
+
+    // Transpose the confidence to match the input shape of `scores` in NMS.
+    // confidence -> {batch, num_boxes, num_classes}
+    // scores -> {batch, num_classes, num_boxes}
+    edsl::Tensor nms_conf = op::transpose(confidence, edsl::make_tuple<int64_t>({0, 2, 1}));
+    // Remove the background class.
+    if (background_label_id > -1) {
+      std::vector<edsl::Tensor> idx_vec;
+      edsl::Tensor idx_first_half = edsl::index({edsl::TensorDim(background_label_id)}, 0);
+      edsl::Tensor idx_second_half =
+          op::slice(edsl::index({edsl::TensorDim(num_classes)}, 0)).add_dim(background_label_id + 1, num_classes);
+      idx_vec.push_back(idx_first_half);
+      idx_vec.push_back(idx_second_half);
+      edsl::Tensor gather_idxs = background_label_id == 0 ? idx_second_half : op::concatenate(idx_vec, 0);
+      nms_conf = edsl::gather(nms_conf, gather_idxs).axis(1);
+    }
+
+    edsl::Tensor iou_threshold = cast(edsl::Tensor{nms_threshold}, DType::FLOAT32);
+    edsl::Tensor score_threshold = cast(edsl::Tensor{confidence_threshold}, DType::FLOAT32);
+    std::vector<edsl::Tensor> result = op::nms(decoded_bboxes, nms_conf, iou_threshold, score_threshold, top_k)
+                                           .soft_nms_sigma(0.0f)
+                                           .center_point_box(false)
+                                           .sort_result_descending(false)
+                                           .box_output_type(DType::INT32)
+                                           .build();
+    edsl::Tensor selected_indices = result[0];
+    auto selected_indices_shape = selected_indices.compute_shape().sizes();
+    edsl::Tensor selected_scores = result[1];
+    edsl::Tensor valid_outputs = result[2];
+
+    edsl::Tensor batch_slice = op::slice(selected_indices).add_dim(0, selected_indices_shape[0]).add_dim(0, 1);
+    edsl::Tensor indices_slice = op::slice(selected_indices).add_dim(0, selected_indices_shape[0]).add_dim(2, 3);
+    std::vector<edsl::Tensor> gather_idxs_vec = {batch_slice, indices_slice};
+    edsl::Tensor gather_idxs = cast(op::concatenate(gather_idxs_vec, 1), DType::INT32);
+    edsl::Tensor out_boxes = edsl::gather(decoded_bboxes, gather_idxs).mode(GatherMode::ND);
+
+    if (clip_after_nms) {
+      out_boxes = op::maximum(cast(edsl::Tensor{0.0f}, DType::FLOAT32),
+                              op::minimum(cast(edsl::Tensor{1.0f}, DType::FLOAT32), out_boxes));
+    }
+
+    edsl::Tensor class_slice = op::slice(selected_indices).add_dim(0, selected_indices_shape[0]).add_dim(1, 2);
```

Comment:
The same issue,  can we use gather?

Reply:
I think we have to handle `class_slice` separately because there's a `background_label_id`  attribute in `DetectionOutput`. Basically it's the index of background class and it (and all those background scores) shouldn't be passed into NMS.

Let's say there are 5 classes total and the index of background class is 2 (usually it's 0 but the specs says it can be any integers). Then what we pass into NMS are scores of class {0, 1, 3, 4} but what NMS returns are scores of class {0, 1, 2, 3}. Hence an offset should be added into {2, 3} in the returned tensor. I'm not sure if there's any better way to do this though.

Reply:
Update the scores of background class to zeros instead of removing the background class.